src/drivers/usb/usb_config.c

```diff
@@ -0,0 +1,191 @@
+	endp->interval = desc->b_interval;
+}
+
+static struct usb_endp *usb_endp_alloc(struct usb_dev *dev, unsigned n,
+		const struct usb_desc_endpoint *endp_desc) {
+	struct usb_endp *endp;
+	struct usb_hcd *hcd = dev->hcd;
+
+	assert(n < USB_DEV_MAX_ENDP);
+	assert(!dev->endpoints[n]);
+
+	endp = pool_alloc(&usb_endps);
+	if (!endp) {
+		return NULL;
+	}
+	endp->dev = dev;
+
+	usb_endp_fill_from_desc(endp, endp_desc);
+	usb_queue_init(&endp->req_queue);
+
+	dev->endpoints[n] = endp;
+
+	if (hcd->ops->endp_hci_alloc) {
+		endp->hci_specific = hcd->ops->endp_hci_alloc(endp);
+		if (!hcd->hci_specific) {
+			pool_free(&usb_endps, endp);
+			return NULL;
+		}
+	}
+
+	return endp;
+}
+
+#if 0
+static void usb_endp_free(struct usb_endp *endp) {
+	struct usb_hcd *hcd = endp->dev->hcd;
+
+	if (hcd->ops->endp_hci_free) {
+		hcd->ops->endp_hci_free(endp, endp->hci_specific);
+	}
+
+	pool_free(&usb_endps, endp);
+}
+#endif
+
+static void usb_parse_and_fill_config(struct usb_dev *dev) {
+	int i;
+	void *cur;
+	struct usb_desc_common_header *other_desc;
+	struct usb_desc_endpoint *endp_desc;
+
+	assert(dev->config_buf);
+
+	cur = dev->config_buf;
+
+	/* Fill interface. */
+	cur += sizeof (struct usb_desc_configuration);
+	memcpy(&dev->iface_desc, cur, sizeof(struct usb_desc_interface));
+	/* b_num_endpoints excludes control endpoint, so add it. */
+	dev->endp_n = dev->iface_desc.b_num_endpoints + 1;
+	cur += sizeof (struct usb_desc_interface);
+
+	/* Skip other descriptors till endpoints. */
+	other_desc = (struct usb_desc_common_header *) cur;
+	while (other_desc->b_desc_type != USB_DESC_TYPE_ENDPOINT) {
+		cur += other_desc->b_length;
+		other_desc = (struct usb_desc_common_header *) cur;
+	}
+
+	/* Fill endpoints. */
+	for (i = 1; i < dev->endp_n; i++) {
+		endp_desc = (struct usb_desc_endpoint *) cur;
+		assert(endp_desc->b_desc_type == USB_DESC_TYPE_ENDPOINT);
+
+		if (NULL == usb_endp_alloc(dev, i, endp_desc)) {
+			panic("%s: failed to alloc endpoint\n",
+					__func__);
+		}
+
+		cur += sizeof (struct usb_desc_endpoint);
+	}
+}
+
+/* Get full device configuration including interfaces and endpoints. */
+int usb_get_configuration(struct usb_dev *dev, unsigned int n) {
+	int len, ret;
+	struct usb_desc_configuration conf;
+
+	/* Check the configuration is not allocated yet. */
+	assert(!dev->config_buf);
+
+	ret = usb_endp_control_wait(dev->endpoints[0],
+		USB_DIR_IN | USB_REQ_TYPE_STANDARD | USB_REQ_RECIP_DEVICE,
+		USB_REQ_GET_DESCRIPTOR,
+		USB_DESC_TYPE_CONFIG << 8,
+		n, sizeof(struct usb_desc_configuration), &conf, 1000);
+	if (ret < 0) {
+		goto err;
+	}
+
+	/* Now get full configuration. */
+	len = conf.w_total_length;
+	dev->config_buf = sysmalloc(len);
+	if (!dev->config_buf) {
+		goto err;
+	}
+
+	ret = usb_endp_control_wait(dev->endpoints[0],
+		USB_DIR_IN | USB_REQ_TYPE_STANDARD | USB_REQ_RECIP_DEVICE,
+		USB_REQ_GET_DESCRIPTOR,
+		USB_DESC_TYPE_CONFIG << 8,
+		n, len, dev->config_buf, 1000);
+	if (ret < 0) {
+		goto err;
```

Comment:
`sysfree(dev->config_buf)` to avoid memory leak

Reply:
Fixed.